### PR TITLE
Vitest prop testing

### PR DESCRIPTION
**Pull request by jessekelly881**
Adds property testing.

[https://www.npmjs.com/package/@fast-check/vitest](https://www.npmjs.com/package/@fast-check/vitest
)

```ts
import { Schema } from "effect"
import { it } from "@effect/vitest"

const realNumber = Schema.Finite.pipe(Schema.nonNaN())

it.prop("symmetry", [realNumber, realNumber], ([a, b]) => a + b === b + a)

it.effect.prop("symmetry", [realNumber, realNumber], ([a, b]) =>
  Effect.gen(function* () {
    yield* Effect.void
    return a + b === b + a
  })
)

it.scoped.prop(
  "should detect the substring",
  { a: Schema.String, b: Schema.String, c: Schema.String },
  ({ a, b, c }) =>
    Effect.gen(function* () {
      yield* Effect.scope
      return (a + b + c).includes(b)
    })
)
```

### CODE REVIEW

**Comment by mikearnaldi** [comment]
why not the following?

```ts
it.effect.prop("should detect the substring", { a: Schema.String, b: Schema.String, c: Schema.String }, ({ a, b, c }) =>
    Effect.gen(function*() {
      return (a + b + c).includes(b)
    })
)
```

**Comment by jessekelly881** [comment]
Yeah, I actually like that more.

**Comment by jessekelly881** [comment]
@mikearnaldi Mind having another look when you get the chance?

**Comment by mikearnaldi** [comment]
Looks good to me! I'll let @tim-smart and @gcanti comment / review

**Comment by jessekelly881** [comment]
There's a weird issue with importing @effect/schema that I'm not quite seeing a solution to.

**Review by tim-smart — APPROVED** [review]

**Comment by tim-smart** [comment]
Whoops wrong PR. I'll review this one shortly.

**Comment by tim-smart** [comment]
Hmm we can't depend on /schema as it isn't part of the core package at this stage.

You can add it to a separate module however.

**Comment by jessekelly881** [comment]
Not even as a devDependency(for every other package besides /vitest)?

**Comment by tim-smart** [comment]
> Not even as a devDependency(for every other package besides /vitest)?

It just can't be used in the /vitest/index.ts module, as /effect dogfoods it.

So you can add it as /vitest/PropertyTesting, but will lose out on the chainable api.

**Comment by mikearnaldi** [comment]
Then I would say let's hold off until we merge schema

**Comment by jessekelly881** [comment]
Hmm. Well this is awkward.. jaja. I've updated it to hopefully fix the issue(hopefully). Still think we should wait?

**Comment by mikearnaldi** [comment]
> Hmm. Well this is awkward.. jaja. I've updated it to hopefully fix the issue. Still think we should wait? What's the timeline for merging schema? If it's w/ 4.0 maybe it's worth merging this version and then updating the api once schema is merged?

Merging schema is not a breaking change for effect so most likely a matter or weeks

**Comment by jessekelly881** [comment]
Gotcha. Makes sense to revert to the original version and wait then. I'll ping this again when Schema gets merged.

**Comment by jessekelly881** [comment]
@tim-smart @mikearnaldi I rebased this against next-minor w/ schema merged into effect. Mind having another look?

File: packages/vitest/src/index.ts
Comment (tim-smart) [inline packages/vitest/src/index.ts:43]: Don't think you need any generics here. Also I don't think you can assign Schema.Never to this.

Comment (jessekelly881) [reply packages/vitest/src/index.ts:43]: Arbitrary doesn't support Schema.Never so I think not being able to assign Schema.Never to it is fine. But, yeah, this can be Schema.All w/o the generics

https://github.com/Effect-TS/effect/blob/b836fc429a9aad36e7a063557d1448ed95398c85/packages/schema/src/Arbitrary.ts#L154

**Review by tim-smart — APPROVED** [review]

**Comment by tim-smart** [comment]
Just need to rebase this again.

**Comment by jessekelly881** [comment]
Done :)

**Comment by tim-smart** [comment]
> Done :)

Rebase, not merge, as there are a bunch of duplicate commits atm.

**Comment by jessekelly881** [comment]
There we go. Had to do some serious git magic to make that work jaja.

File: packages/vitest/src/internal.ts
Comment (adrian-gierakowski) [reply packages/vitest/src/internal.ts:20]: https://github.com/Effect-TS/effect/pull/4230

**Pull request merged**
